btf/types.go

```diff
@@ -67,13 +71,73 @@ func (ts types) ByID(id TypeID) (Type, error) {
 	return ts[id], nil
 }
 
+type Skipper func(Type) bool
+
+type Comparer func(Type, Type) bool
+
+// SkipQualifierAndTypedef skips qualifiers and Typedefs.
+var SkipQualifierAndTypedef = func(t Type) bool {
+	switch t.(type) {
+	case qualifier, *Typedef:
+		return true
+	}
+	return false
+}
+
+// CompareTypeName returns true if a and b's names correspond.
+var CompareTypeName = func(a, b Type) bool {
+	return a.TypeName() == b.TypeName()
+}
+
+// Equal compares Type a to Type b. By default, only binary equivalency is
+// checked, ignoring type names.
+//
+// cmp allows providing extra comparison rules. It will be evaluated after
+// checking a and b's binary equivalency at each step of the type hierarchy.
+// skip allows individual Types to be omitted from comparisons, and is called
+// before checking a and b's binary equivalency.
+func Equal(a, b Type, cmp Comparer, skip Skipper) error {
+	if a == nil && b == nil {
+		return nil
+	}
+	if a == nil || b == nil {
+		return errors.New("one of a or b are nil")
+	}
+
+	at, bt := postorderTraversal(a, nil), postorderTraversal(b, nil)
+	for at.Next() && bt.Next() {
+		if skip != nil {
```

Comment:
Allowing skipping like this is dangerous, I think. postorderTraversal flattens a graph into a list, so we lose the shape of the graph. This is okay (I think, need to convince myself of that again) as long as we always proceed in lockstep, consuming one type from each list. This skipping breaks that invariant so I think you can end up comparing types from different branches of a type.

Reply:
All valid concerns indeed, and of course skips need to be used with care, that's somewhat self-explanatory.

The hierarchies are indeed flattened, but a and b are flattened the exact same way, and the same skip() is executed on both of them. At some point the comparison will converge on a complex (struct/union/func) parent node. If either side skipped a whole branch, one side will still be evaluating the branch while the other hits the parent, and a mismatch will occur.

I don't think this is a huge concern as it will flag a mismatch at some point, but it does call into question the usefulness of skipping anything other than qualifiers and typedefs. At worst the output will be inaccurate.

Reply:
> The hierarchies are indeed flattened, but a and b are flattened the exact same way [...]. If either side skipped a whole branch, one side will still be evaluating the branch while the other hits the parent, and a mismatch will occur.

I had something like these types in mind. Would this mechanism still work for them? I'm kinda sceptical (and now think that the CO-RE logic is buggy because of this).

```go
struct {
  a struct{}
  b int
  c int
}

struct {
  a struct {
    b int
    c int
  }
}
```

> and the same skip() is executed on both of them.

That assumes that skip is a pure function. I've already been bitten by that not being true when using `postorderTraversal` in the Spec.Add PR, and I wrote that code!

> but it does call into question the usefulness of skipping anything other than qualifiers and typedefs.

Yes, indeed. However, I seem to remember a case where I wanted to skip only qualifiers and not typedefs though (GoFormatter maybe?) That implies that just a `skip bool` doesn't cut it.

Reply:
> That implies that just a `skip bool` doesn't cut it.

Totally, just being devil's advocate. Equal() should always take some form of function for maximum flexibility and composability.

Reply:
> Would this mechanism still work for them? [..example..]

With the current PR, this should flatten both hierarchies into [c, b, a, <parent struct>], and both legs will visit them in the same order.

c and b will indeed match, even though they're not at the same level in the hierarchy. a will not match because the left side has 0 members and the right side has two members. Likewise, `<parent struct>` has 3 members on the left and 1 member on the right, so won't match either. See what I mean?